src/utils/dom-fingerprint.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+
+/**
+ * Gets a string showing the index of an element within
+ * the children of its parent, counting only nodes with the same tag.
+ * Stop at 25, just to have a limit.
+ * @param {?Element} element DOM node to get index of.
+ * @return {string} '.<index>' or ''.
+ */
+function indexWithinParent(element) {
+  if (element && element.nodeName && element.parentElement) {
+    const elementParent = element.parentElement;
+    const nodeName = element.nodeName.toString().toLowerCase();
+    // Find my index within my parent's children
+    const children = elementParent.childNodes;
+    // Choose a limit that we hope will allow getting to 25
+    // matching nodes.
+    const limit = Math.min(children.length, 100);
+    let matchingNodeCount = 0;
+    for (let i = 0; i < limit && matchingNodeCount < 25; i++) {
+      const child = children[i];
+      // Some browsers treat childNodes differently.
+      // So we'll only count nodes with the same tag.
+      if (child.nodeName &&
+          child.nodeName.toString().toLowerCase() === nodeName) {
```

Comment:
We can avoid these `#toLowerCase`s by not calling it on the `element`'s `nodeName`.